keps/sig-network/1880-services-ip-ranges/README.md

```diff
@@ -0,0 +1,1015 @@
+
+#### The special "default" ServiceIPRanges
+
+There has to be always one default ServiceIPRange containing the `kubernetes.default` Service.
+
+If the cluster is dual-stack, there has to be also one default ServiceIPRange per IP family.
+
+In order to ensure the flags are covered by a range, the controller will handle the case of the
+ServiceIPRange being deleted forcefully by recreating it. We allow overlapping ranges for this
+reason, and the default ServiceIPRange is denoted by a wellknown label, not a special name.
+
+The controller will also handle the case where the well-known label is removed, re-adding it.
+
+#### Service IP Allocation
+
+When a a new Service is going to be created and already has defined the ClusterIPs, the allocator will
+just try to create the corresponding IPAddress objects, any error creating the IPAddress object will
+cause an error on the Service creation. The allocator will also set the reference to the Service on the
+IPAddress object.
+
+The new allocator will have a local list with all the IP ranges and IP already allocated, so it will
+have to check just for one free IP in any of these ranges and try to create the object.
+There can be races with 2 allocators trying to allocate the same IP, but the storage guarantees the consistency
+and the first will win, meanwhile the other will have to retry.
+
+Another racy situation is when the allocator is full and an IPAddress is deleted but the allocator takes
+some time to receive the notification. One solution can be to perform a List to obtain current state, but
+it will be simpler to just fail to create the Service and ask the user to try to create the Service again.
+
+If the apiserver crashes during a Service create operation, the IPAddress is allocated but the
+Service is not created, the IPaddress will be orphan. To avoid leaks, a controller will use the
+`metadata.creationTimestamp` field to detect orphan objects and delete them.
+
+There is going to be a controller to avoid leaking resources:
+
+- checking that the corresponding parentReference on the IPAddress match the corresponding Service
```

Comment:
Regarding UID - I spoke with @deads2k and he's OK to move UID init earlier, so if we need it, we can use it.

Reply:
> Now we have 2 IPs that seem valid - they both refer to "x", and "x" exists. As long as the IP GC controller verifies that `IP.ref = x" and "x.spec.clusterIPs[] includes that IP" I think we're OK without UID. Agree?

right, we have the information on both the service and the IP to avoid the need of UIDs,

Reply:
Coming back to this - if we have UID, we can work with any parent (e.g. gateway)

Reply:
yep, UID solves it and it seems we have it covered https://github.com/kubernetes/kubernetes/pull/110646